docs/spec_draft.md

```diff
@@ -825,6 +826,247 @@ Produces an `output` tensor from a constant `value`.
+
+Given a `result` index `result_index`, the following pseudo-code finds all the
+`lhs` and `rhs` index pairs, `lhs_full_index` and `rhs_full_index` resp.,
+responsible in computing the value at `result_index` using dot product.
+
+```python
+value = 0;
+foreach rhs spatial index [k0, k1, ...]
+  ##
+  ## Compute the components in lhs_full_index at spatial dimensions.
+  ##
+  i0' = result_index[output_spatial_dimension[0]] * window_strides[0] - padding[0][0] + k0 * rhs_dilation[0]
+  i0 = base_dilations[0] > 1 ? i0' / base_dilations[0] : i0'
+
+  i1' = result_index[output_spatial_dimension[1]] * window_strides[1] - padding[1][0] + k1 * rhs_dilation[1]
+  i1 = base_dilations[1] > 1 ? i1' / base_dilations[1] : i1'
+  ... omiting similar logic for other spatial dimensions ...
+
+  if i0' % lhs_dilation[0] != 0 or i0 not in bounds of dim(lhs, 0), then skip [k0, k1]
+  if i1' % lhs_dilation[1] != 0 or i1 not in bounds of dim(lhs, 1), then skip [k0, k1]
+  ... omiting similar logic for other spatial dimensions ...
+
+  lhs_full_index[input_spatial_dimension] = [i0, i1, ...]
+
+  ##
+  ## Compute the components in rhs_full_index at spatial dimensions.
+  ##
+  rhs_full_index[kernel_spatial_dimension] = [
+    window_reversal[0] ? dim(rhs, kernel_spatial_dimension[0]) - 1 - k0 : k0,
+    window_reversal[1] ? dim(rhs, kernel_spatial_dimension[1]) - 1 - k1 : k1
+    ... omiting similar logic for other spatial dimensions ...
+  ]
+
+  input_feature_group_size = dim(lhs, input_feature_dimension) / feature_group_count.
+  foreach rhs_iz in [0, input_feature_group_size)
+    ##
+    ## Compute the components in lhs_full_index at input_batch_dimension  and input_feature_dimension.
+    ##
+    lhs_full_index[input_batch_dimension] = result_index[output_batch_dimension] + ((batch_group_index * batch_group_size) % input_batch_size)
+      where
+        batch_group_size = input_batch_size / batch_group_count
+        input_batch_size = dim(lhs, input_batch_dimension)
+        batch_group_index = result_index[output_feature_dimension] / depthwise_multiplier
+        depthwise_multiplier = batch_group_count > 1 ? dim(rhs, kernel_output_feature_dimension) / batch_group_count : 1
+
+    lhs_full_index[input_feature_dimension] = rhs_iz + feature_group_index * input_feature_group_size
+      where
+        feature_group_index = result_index[output_feature_dimension] / output_feature_group_size
+        output_feature_group_size = dim(rhs, kernel_output_feature_dimension) / feature_group_count
+
+    ##
+    ## Compute the components in rhs_full_index at kernel_input_feature_dimension and kernel_output_feature_dimension.
+    ##
+    rhs_full_index[kernel_input_feature_dimension] = rhs_iz
+
+    rhs_full_index[kernel_output_feature_dimension] = result_index[output_feature_dimension]
+
+    ##
+    ## Accumulate value to be stored in result at 'result_index'.
+    ##
+    value += rhs[rhs_full_index] * lhs[lhs_full_index]
+
+  result[result_index] = value
+```
```

Comment:
Here's a formalization:

```
More formally, we start with reframing the inputs to the operation in terms 
of `lhs` in order to be able to express windows of `lhs`:
  * `lhs_window_dimensions = lhs_shape(dim(lhs, input_batch_dimension), rhs_to_lhs(dim(rhs, kernel_spatial_dimensions)), dim(lhs, input_feature_dimension))`.
  * `lhs_window_strides = lhs_shape(1, ascending_to_lhs(window_strides), 1)`.
  * `lhs_padding = lhs_shape([0, 0], ascending_to_lhs(padding), [0, 0])`.
  * `lhs_base_dilations = lhs_shape(1, ascending_to_lhs(lhs_dilation), 1)`.
  * `lhs_window_dilations = lhs_shape(1, ascending_to_lhs(rhs_dilation), 1)`.

This reframing uses the following helper functions:
  *  `lhs_shape(n, hw, c) = transpose([n] + hw + [c], [input_batch_dimension] + input_spatial_dimensions + [input_feature_dimension])`.
  *  `rhs_shape(hw, i, o) = transpose(hw + [i, o], kernel_spatial_dimensions, [kernel_input_feature_dimension, kernel_output_feature_dimension])`.
  *  `result_shape(n1, hw, c1) = transpose([n] + hw + [c], [output_batch_dimension] + output_spatial_dimensions + [output_feature_dimension])`.
  *  `source_to_target(hw)` transposes a list that is ordered according to 
     `source_spatial_dimensions`, so that it becomes ordered according to 
     `target_spatial_dimensions`.

If `feature_group_count = 1` and `batch_group_count = 1`, then for all 
`result_spatial_index` in the index space of `result_spatial_dimensions`, 
`result[result_shape(:, result_spatial_index, :)] = dot_product` where:
  * `padded_lhs = pad(lhs, 0, lhs_padding[:, 0], lhs_padding[:, 1], lhs_base_dilations)`.
  * `lhs_window_start = lhs_shape(0, result_to_lhs(result_spatial_index), 0) * lhs_window_strides`.
  * `lhs_window = slice(padded_lhs, lhs_window_start, lhs_window_start + lhs_window_dimensions, lhs_window_dilations)`.
  * `dot_product = dot_general(lhs_window, rhs, 
      lhs_batching_dimensions=[],
      lhs_contracting_dimensions=input_spatial_dimensions + [input_feature_dimension],
      rhs_batching_dimensions=[],
      rhs_contracting_dimensions=kernel_spatial_dimensions + [kernel_input_feature_dimension])`.

If `feature_group_count > 1`:
  * `lhses = split(lhs, feature_group_count, input_feature_dimension)`.
  * `rhses = split(rhs, feature_group_count, kernel_output_feature_dimension)`.
  * `results[:] = convolution(lhses[:], rhses[:], ..., feature_group_count=1)`.
  * `result = concatenate(results, output_feature_dimension)`.

If `batch_group_count > 1`:
  * `lhses = split(lhs, batch_group_count, input_batch_dimension)`.
  * `rhses = split(rhs, batch_group_count, kernel_output_feature_dimension)`.
  * `results[:] = convolution(lhses[:], rhses[:], ..., batch_group_count=1)`.
  * `result = concatenate(results, output_feature_dimension)`.

Where `split(tensor, num_splits, dimension)` splits a `tensor` into `num_splits`
tensors of the same shape, so that `concatenate(results, dimension)` produces
back the same `tensor`.
```

Reply:
Changes:

1. Dropping `ascending_to_lhs` in

```
`lhs_window_strides = lhs_shape(1, ascending_to_lhs(window_strides), 1)`.
`lhs_shape(n, hw, c) = permute([n] + hw + [c], [input_batch_dimension] + input_spatial_dimensions + [input_feature_dimension])`.
```

Problem: double permute, `ascending_to_lhs` does the first permute and then `lhs_shape` does the second, which we may not want.

The reasoning can be further explained using an example which is intentionally made
unconventional but is still legal:

Suppose `lhs_shape : [a,b,c,d]` with
  input_batch_dimension: 1
  input_feature_dimension: 2
  input_spatial_dimensions: [3, 0]

Suppose `rhs_shape : [p,q, r, s]` with
  kernel_input_feature_dimension: 2
  kernel_output_feature_dimension: 3
  kernel_spatial_dimensions: [1,0]

Suppose window_strides: [X, Y]
which essentially says: the stride value for the 0th spatial axis is `X` and that for 1st spatial axis is `Y`.

The idea behind doing permutation while creating `lhs_window_strides` is to match its shape with that of `lhs`.

Ideally this should be [Y, 1, 1, X]
The reason is: As per `input_spatial_dimensions [3, 0]`, the 0th spatial dimension
corresponds to 3rd axis in `lhs` and 1st spatial dimension corresponds to 0th axis of `lhs`.

To get that we can  do
```
`lhs_window_strides = lhs_shape(1, window_strides, 1)`
`lhs_shape(n, hw, c) = permute([n] + hw + [c], [input_batch_dimension] + input_spatial_dimensions + [input_feature_dimension])`.
                     = permute([1,X,Y,1], [1,3,0,2]) = [Y, 1,1, X].
```

2. Drop `rhs_to_lhs` from
```
`lhs_window_dimensions = lhs_shape(dim(lhs, input_batch_dimension), rhs_to_lhs(dim(rhs, kernel_spatial_dimensions)), dim(lhs, input_feature_dimension))`.
```

Using the running example,
ideally we want `lhs_window_dimensions` to be `[p, b, c, q]`.

That is because as per ` kernel_spatial_dimensions: [1,0]` and `rhs_shape : [p,q,r,s]`
The dimension size of 0th axis is `q` and that of 1st axis is `p`.

To get that, we can do
```
`lhs_window_dimensions = lhs_shape(dim(lhs, input_batch_dimension), dim(rhs, kernel_spatial_dimensions), dim(lhs, input_feature_dimension))`.
                       = permute([b, q, p, c], [1,3,0,2])
                       = [p, b, c, q]
```


3. Dropped `source_to_target` as we do not need it.

4. `result_spatial_dimensions` to `output_spatial_dimensions`.

5.  `index space of output_spatial_dimensions`: I think you meant `index space
    of dim(result, output_spatial_dimensions)`?

Reply:
6. Removed the unused macros like `rhs_shape` and `source_to_target`

Reply:
These are some very nice simplifications! They helped me formalize DotGeneralOp as well.